slick/src/main/scala/slick/basic/BasicBackend.scala

```diff
@@ -309,45 +309,47 @@ trait BasicBackend { self =>
+                if(ctx.cancelled) {
+                  if(ctx.deferredError ne null) throw ctx.deferredError
+                  if(state ne null) { // streaming cancelled before finishing
+                    val oldState = state
+                    state = null
+                    a.cancelStream(ctx, oldState)
+                  }
+                } else if(realDemand > 0 || (state eq null)) {
                   val oldState = state
                   state = null
-                  a.cancelStream(ctx, oldState)
+                  state = a.emitStream(ctx, realDemand, oldState)
                 }
-              } else if((realDemand > 0 || (state eq null))) {
-                val oldState = state
-                state = null
-                state = a.emitStream(ctx, realDemand, oldState)
-              }
-              if(state eq null) { // streaming finished and cleaned up
+                if(state eq null) { // streaming finished and cleaned up
+                  releaseSession(ctx, true)
+                  ctx.streamingResultPromise.trySuccess(null)
+                }
+              } catch { case NonFatal(ex) =>
+                if(state ne null) try a.cancelStream(ctx, state) catch ignoreFollowOnError
                 releaseSession(ctx, true)
-                ctx.streamingResultPromise.trySuccess(null)
+                throw ex
               }
-            } catch { case NonFatal(ex) =>
-              if(state ne null) try a.cancelStream(ctx, state) catch ignoreFollowOnError
-              releaseSession(ctx, true)
-              throw ex
-            } finally {
-              ctx.streamState = state
-              if (!ctx.isPinned && ctx.priority(continuation) != WithConnection) connectionReleased = true
-              ctx.sync = 0
-            }
-            if(debug) {
-              if(state eq null) streamLogger.debug(s"Sent up to ${str(realDemand)} elements - Stream " + (if(ctx.cancelled) "cancelled" else "completely delivered"))
-              else streamLogger.debug(s"Sent ${str(realDemand)} elements, more available - Performing atomic state transition")
-            }
-            demand = ctx.delivered(demand)
-            realDemand = if(demand < 0) demand - Long.MinValue else demand
-          } while ((state ne null) && realDemand > 0)
+              if(debug) {
+                if(state eq null) streamLogger.debug(s"Sent up to ${str(realDemand)} elements - Stream " + (if(ctx.cancelled) "cancelled" else "completely delivered"))
+                else streamLogger.debug(s"Sent ${str(realDemand)} elements, more available - Performing atomic state transition")
+              }
+              demand = ctx.delivered(demand)
+              realDemand = if(demand < 0) demand - Long.MinValue else demand
+            } while ((state ne null) && realDemand > 0)
+          } finally {
```

Comment:
I'm a bit worried about moving the `finally` block outside of the while loop, because it might change the behaviour, because the context is now only updated when the streaming is finished instead of after each batch delivery. But I don't know that part of the code that well.

Maybe it's better to leave the `finally` block inside the while loop and also a duplicate it outside to protect the `acquireSession` method throwing? WDYT?

Reply:
I agree. I think this could be  reviewed so the `while` loop (and related logs statements) was extracted to a separate private method too. It'd be clearer which `try` goes with which `catch/finally`.

Reply:
ok, I will review this part. I think it's fair to say that the outer `finally` should only care about marking the runnable as "released" so the thread is liberated. The state should stay inside the loop. State is only gets updated inside the loop and after acquiring a session.